docs/clients/dotnet/21.2/migration-to-gRPC.md

```diff
@@ -0,0 +1,630 @@
+            .Select(ToEventData)
+            .ToArray();
+
+        return tcpConnection.AppendToStreamAsync(
+            streamName,
+            version,
+            preparedEvents
+        );
+
+        static EventData ToEventData(object @event) =>
+            new EventData(
+                Guid.NewGuid(),
+                TypeMapper.GetTypeName(@event.GetType()),
+                true,
+                Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(data))
+            );
+    }
+
+    public Task AppendEvents(
+        string streamName,
+        params object[] events
+    )
+        => AppendEvents(streamName, ExpectedVersion.Any, events);
+
+    public async Task<IEnumerable<object>> LoadEvents(string stream)
+    {
+        const int pageSize = 4096;
+
+        var start  = 0;
+        var events = new List<object>();
+
+        do
+        {
+            var page = await tcpConnection.ReadStreamEventsForwardAsync(
+                stream, start, pageSize, true
+            );
+
+            if (page.Status == SliceReadStatus.StreamNotFound)
+                throw new ArgumentOutOfRangeException(
+                    nameof(stream), $"Stream '{stream}' was not found"
+                );
+
+            events.AddRange(
+                page.Events.Select(Deserialize)
+            );
+            if (page.IsEndOfStream) break;
+
+            start += pageSize;
+        } while (true);
+
+        return events;
+
+        static object Deserialize(this ResolvedEvent resolvedEvent)
+        {
+            var dataType = TypeMapper.GetType(resolvedEvent.Event.EventType);
+            var jsonData = Encoding.UTF8.GetString(resolvedEvent.Event.Data);
+            var data = JsonConvert.DeserializeObject(jsonData, dataType);
+            return data;
+        }
+    }
+
+    public async Task<bool> StreamExists(string stream)
+    {
+        var result = await tcpConnection.ReadEventAsync(stream, 1, false);
+        return result.Status != EventReadStatus.NoStream;
+    }
+}
+```
+
+## Differences in connection management
+
+Both the TCP and gRPC clients are managing reconnections. That's the reason why it should be registered as a single instance in the application. 
+
+The TCP client requires calling the `ConnectAsync` method at least once to initiate the connection. That wasn't ideal if you wanted to inject an already set up connection, as you either had to call it asynchronously or risk deadlocks.
+
+```csharp
+private async Task<IEventStoreConnection> GetEventStoreConnection(string connectionString)
+{
+    var tcpConnection = EventStoreConnection.Create(connectionString);
+    await tcpConnection.ConnectAsync();
+    return tcpConnection;
+}
+```
+
+As TCP client connection logic is not thread-safe (`ConnectAsync` method), other operations are thread-safe. Because of that, you have to ensure that you won't have a race condition. You also should consider handling `Closed` events to manage reconnection if it is closed. 
```

Comment:
```suggestion
The TCP client connection logic is not thread-safe (`ConnectAsync` method), however the other operations are thread-safe. Because of that, you have to ensure that you won't have a race condition. You also should consider handling `Closed` events to manage reconnection if it is closed. 
```